File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class comperator{
    public:
        bool operator()(ListNode* &a,ListNode* &b){
            return a->val > b->val;
        }
};
class Solution {
    
    void insertAtTail(ListNode* &head,ListNode* &tail,ListNode* &nextNode){
        if(head==NULL and tail==NULL){
            head = nextNode;
            tail = nextNode;
        }else{
            tail->next = nextNode;
            tail = tail->next;
        }
    }
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*,vector<ListNode*>,comperator> pq;

        for(ListNode* head: lists){
            if(head!=NULL) pq.push(head);
        }

        ListNode* head = NULL;
        ListNode* tail = NULL;
        while(!pq.empty()){
            ListNode* nextNode = pq.top(); pq.pop();

            insertAtTail(head,tail,nextNode);
            if(nextNode->next!=NULL) pq.push(nextNode->next);
        }
        return head;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
import heapq
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertAtTail(self,head,tail,nextNode) ->Optional[ListNode]:
        if(head is None and tail is None):
            head = nextNode
            tail = nextNode
        else:
            tail.next = nextNode
            tail = tail.next
        return head,tail
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        minHeap = []
        for i in range(len(lists)):
            if(lists[i]!=None):
                heapq.heappush(minHeap,(lists[i].val,i,lists[i]))

        head = None
        tail = None

        while(len(minHeap)>0):
            val,i,nextNode = heapq.heappop(minHeap)

            head,tail = self.insertAtTail(head,tail,nextNode)

            if(nextNode.next):
                heapq.heappush(minHeap,(nextNode.next.val,i,nextNode.next))

        return head

            
        